File: tools/migrate/providers/zep.js
const fs = require('fs'),
  p = require('path');
class Z {
  constructor(c) {
    this.c = c;
    this.k = c.k;
    this.u = c.u || 'https://api.getzep.com';
    this.rl = c.rl || 1;
    this.h = {
      Authorization: `Bearer ${this.k}`,
      'Content-Type': 'application/json',
    };
    this.d = 1000 / this.rl;
    this.l = 0;
  }
  async w() {
    const n = Date.now(),
      e = this.l + this.d - n;
    if (e > 0) await new Promise((r) => setTimeout(r, e));
    this.l = Date.now();
  }
  async conn() {
    try {
      await this.w();
      const r = await this.f('/api/v2/sessions?limit=1');
      return { ok: true, ses: r.total || 0 };
    } catch (e) {
      throw new Error(`Zep conn fail: ${e.message}`);
    }
  }
  async exp() {
    const o = p.join(this.c.o, 'zep_export.jsonl'),
      w = fs.createWriteStream(o);
    console.log('[ZEP] Listing sessions...');
    const ss = await this.fss();
    console.log(`[ZEP] Found ${ss.length} sessions`);
    let tm = 0,
      tc = 0;
    for (let i = 0; i < ss.length; i++) {
      const s = ss[i];
      tc++;
      if (tc % 100 === 0)
        console.log(`[ZEP] Progress: ${tc}/${ss.length} sessions`);
      const mm = await this.fsm(s.session_id);
      for (const m of mm) {
        w.write(JSON.stringify(this.n(m, s)) + '\n');
        tm++;
      }
    }
    w.end();
    console.log(`[ZEP] Exported ${tm} memories from ${tc} sessions`);
    return o;
  }
  async fss() {
    const ss = [];
    let pg = 1,
      l = 100;
    while (true) {
      await this.w();
      const r = await this.f(`/api/v2/sessions?page=${pg}&limit=${l}`),
        b = r.sessions || [];
      if (!b.length) break;
      ss.push(...b);
      pg++;
      if (b.length < l) break;
    }
    return ss;
  }
  async fsm(sid) {
    try {
      await this.w();
      const r = await this.f(`/api/v2/sessions/${sid}/memory`);
      return r.messages || r.memories || [];
    } catch (e) {
      console.warn(`[ZEP] Warn: session ${sid.slice(0, 8)} no memories`);
      return [];
    }
  }
  n(m, s) {
    return {
      id: m.uuid || m.id || `${s.session_id}_${Date.now()}`,
      uid: s.user_id || s.session_id || 'default',
      c: m.content || m.text || '',
      t: m.metadata?.tags || [],
      meta: {
        p: 'zep',
        sid: s.session_id,
        r: m.role,
        tc: m.token_count,
        om: m.metadata || {},
      },
      ca: m.created_at ? new Date(m.created_at).getTime() : Date.now(),
      ls: m.updated_at ? new Date(m.updated_at).getTime() : Date.now(),
      e: m.embedding || null,
    };
  }
  async f(ep) {
    const u = `${this.u}${ep}`,
      r = await fetch(u, { method: 'GET', headers: this.h });
    if (!r.ok) {
      if (r.status === 429) {
        const rt = r.headers.get('retry-after') || 60;
        console.warn(`[ZEP] Rate limit, waiting ${rt}s...`);
        await new Promise((x) => setTimeout(x, rt * 1000));
        return this.f(ep);
      }
      throw new Error(`HTTP ${r.status}`);
    }
    return await r.json();
  }
}
module.exports = Z;

File: tools/migrate/providers/zep.py
from typing import Dict, Any, AsyncGenerator, List
from .base import BaseProvider
from ..schemas import MigrationRecord
from ..utils import logger

class ZepProvider(BaseProvider):
    def __init__(self, config):
        super().__init__(config)
        self.base_url = config.source_url or "https://api.getzep.com"
        self.headers = {
            "Authorization": f"Bearer {config.api_key}",
            "Content-Type": "application/json"
        }

    async def connect(self) -> Dict[str, Any]:
        url = f"{self.base_url}/api/v2/sessions?limit=1"
        try:
            data = await self._get(url, headers=self.headers)
            return {"ok": True, "sessions": data.get("total", 0)}
        except Exception as e:
            raise Exception(f"Zep connection failed: {e}")

    async def export(self) -> AsyncGenerator[MigrationRecord, None]:
        try:
            sessions = await self._fetch_all_sessions()
            logger.info(f"[ZEP] Found {len(sessions)} sessions")
            
            for i, session in enumerate(sessions):
                if i % 100 == 0:
                    logger.info(f"[ZEP] Processing session {i}/{len(sessions)}")
                
                memories = await self._fetch_session_memories(session["session_id"])
                for mem in memories:
                    yield self._transform(mem, session)
        except Exception as e:
            logger.error(f"[ZEP] Export failed: {e}")
            raise

    async def _fetch_all_sessions(self) -> List[Dict]:
        sessions = []
        page = 1
        limit = 100
        while True:
            url = f"{self.base_url}/api/v2/sessions?page={page}&limit={limit}"
            data = await self._get(url, headers=self.headers)
            batch = data.get("sessions", [])
            if not batch:
                break
            sessions.extend(batch)
            page += 1
            if len(batch) < limit:
                break
        return sessions

    async def _fetch_session_memories(self, session_id: str) -> List[Dict]:
        url = f"{self.base_url}/api/v2/sessions/{session_id}/memory"
        try:
            data = await self._get(url, headers=self.headers)
            return data.get("messages", []) or data.get("memories", [])
        except Exception:
            return []

    def _transform(self, m: Dict, s: Dict) -> MigrationRecord:
        return MigrationRecord(
            id=m.get("uuid") or m.get("id") or f"{s['session_id']}_{m.get('created_at', '')}",
            uid=s.get("user_id") or s.get("session_id") or "default",
            content=m.get("content") or m.get("text") or "",
            tags=m.get("metadata", {}).get("tags", []),
            metadata={
                "provider": "zep",
                "session_id": s.get("session_id"),
                "role": m.get("role"),
                "original_metadata": m.get("metadata", {})
            },
            created_at=m.get("created_at"), # String to pass through or parse? Schema said int.
            # Schema says int (timestamp). Need to parse if string.
            # For simplicity let's store what we get or parse if easy. JS version used Date.parse
        )
